controlplane/kubeadm/internal/cluster.go

```diff
@@ -180,16 +181,15 @@ func (m *Management) getApiServerEtcdClientCert(ctx context.Context, clusterKey
 	return tls.X509KeyPair(crtData, keyData)
 }
 
-type healthCheck func(context.Context) (HealthCheckResult, error)
-
 // HealthCheck will run a generic health check function and report any errors discovered.
 // In addition to the health check, it also ensures there is a 1;1 match between nodes and machines.
-func (m *Management) healthCheck(ctx context.Context, check healthCheck, clusterKey client.ObjectKey) error {
+// To have access to the owned control-plane machines during health checks, need to pass owningMachines here.
+func (m *Management) healthCheck(controlPlane *ControlPlane, nodeChecks HealthCheckResult, clusterKey client.ObjectKey) error {
 	var errorList []error
-	nodeChecks, err := check(ctx)
-	if err != nil {
-		errorList = append(errorList, err)
-	}
+	kcpMachines := controlPlane.Machines.UnsortedList()
+	// Make sure Cluster API is aware of all the nodes.
+
+	// TODO: If any node has a failure, healthCheck fails. This may be too strict for the health check of HA clusters (Missing a single etcd pod does not indicate a problem)...
```

Comment:
KCP health checks are specifically designed to ensure a kubeadm generated control-plane is stable before 
proceeding with KCP actions like scale-up.

This could be seen as a strict, but given that scaling up could increase the number of members required for etcd quorum, this can lead to problems if we allow scaling up when there existing etcd members (it is the same reason behind the decision to scale in when doing remediations). 

If we are willing to change this mechanism, which is responsible for the "slowness" of KCP, I would suggest opening an issue (instead of a TODO) so we can have a proper discussion about pro and cons.

Reply:
Got it. Removed this todo.